Valgrind reports errors on run-and-quit with Royal92

I've been dipping my toe into the investigation of memory leaks because running verify.ll on my database of 10,000 or so INDIs is hogging 15 GB of memory and taking an hour and a half to complete. I'd also noticed that once verify.ll had run to completion llines still held on to about 500 MB of memory.

To investigate the problem without risking my live database on my Mac, I've been using an Ubuntu 20.04 virtual machine, and ran the following and then hit "q" as soon as the menu came up...

valgrind -s --leak-check=full --show-leak-kinds=all llines Royal92

This generated an 800+ line report, the majority of which is probably unimportant. On the other hand, the four errors below are alarming (to me, at least!).



...
==113525== LEAK SUMMARY:
==113525==    definitely lost: 0 bytes in 0 blocks
==113525==    indirectly lost: 0 bytes in 0 blocks
==113525==      possibly lost: 0 bytes in 0 blocks
==113525==    still reachable: 592,318 bytes in 538 blocks
==113525==         suppressed: 0 bytes in 0 blocks
==113525== 
==113525== ERROR SUMMARY: 4 errors from 4 contexts (suppressed: 0 from 0)
==113525== 
==113525== 1 errors in context 1 of 4:
==113525== Invalid read of size 8
==113525==    at 0x11AB44: save_nkey_list (browse.c:1657)
==113525==    by 0x116945: main (main.c:421)
==113525==  Address 0x4d52e98 is 24 bytes before a block of size 640 alloc'd
==113525==    at 0x483C855: malloc (vg_replace_malloc.c:380)
==113525==    by 0x17DC35: __allocate (memalloc.c:72)
==113525==    by 0x11A803: init_hist (browse.c:1547)
==113525==    by 0x11EDA5: load_hist_lists (browse.c:1519)
==113525==    by 0x11EDA5: init_browse_module (browse.c:2051)
==113525==    by 0x1168F4: main (main.c:404)
==113525== 
==113525== 
==113525== 1 errors in context 2 of 4:
==113525== Invalid read of size 1
==113525==    at 0x11AB1B: save_nkey_list (browse.c:1655)
==113525==    by 0x116945: main (main.c:421)
==113525==  Address 0x4d52e90 is 32 bytes before a block of size 640 in arena "client"
==113525== 
==113525== 
==113525== 1 errors in context 3 of 4:
==113525== Invalid read of size 8
==113525==    at 0x11AB44: save_nkey_list (browse.c:1657)
==113525==    by 0x11EE32: save_hist_lists (browse.c:1534)
==113525==    by 0x11EE32: term_browse_module (browse.c:2061)
==113525==    by 0x116945: main (main.c:421)
==113525==  Address 0x4d52bd8 is 24 bytes before a block of size 640 alloc'd
==113525==    at 0x483C855: malloc (vg_replace_malloc.c:380)
==113525==    by 0x17DC35: __allocate (memalloc.c:72)
==113525==    by 0x11A803: init_hist (browse.c:1547)
==113525==    by 0x11ED96: load_hist_lists (browse.c:1518)
==113525==    by 0x11ED96: init_browse_module (browse.c:2051)
==113525==    by 0x1168F4: main (main.c:404)
==113525== 
==113525== 
==113525== 1 errors in context 4 of 4:
==113525== Invalid read of size 1
==113525==    at 0x11AB1B: save_nkey_list (browse.c:1655)
==113525==    by 0x11EE32: save_hist_lists (browse.c:1534)
==113525==    by 0x11EE32: term_browse_module (browse.c:2061)
==113525==    by 0x116945: main (main.c:421)
==113525==  Address 0x4d52bd0 is 32 bytes before a block of size 640 in arena "client"
==113525== 
==113525== ERROR SUMMARY: 4 errors from 4 contexts (suppressed: 0 from 0)
parallels@parallels-Parallels-Virtual-Platform:~$

The message on running "llines Royal92" is "Bad history count". I haven't edited Royal92, just imported the Royal92.ged file.

I downloaded the current version of lldump.c, but it failed at the make stage with the following:

gcc  -g -O2 -W -Wall -Wcast-align -Wmissing-declarations -Wmissing-prototypes -Wreturn-type -Wstrict-prototypes -pedantic -g   -o dbverify dbverify.o wprintf.o ../gedlib/libgedcom.a ../btree/libbtree.a ../stdlib/libstdlib.a ../arch/libarch.a  -lncursesw -lm 
gcc -DLOCALEDIR=\"/usr/local/share/locale\" -DHAVE_CONFIG_H -I. -I../..  -I../.. -I../../src/hdrs   -g -O2 -W -Wall -Wcast-align -Wmissing-declarations -Wmissing-prototypes -Wreturn-type -Wstrict-prototypes -pedantic -g -MT lldump.o -MD -MP -MF .deps/lldump.Tpo -c -o lldump.o lldump.c
lldump.c: In function ‘print_index’:
lldump.c:442:10: error: ‘FMT_INT32_HEX_06’ undeclared (first use in this function); did you mean ‘FMT_INT32_HEX’?
  442 |   printf(FMT_INT32_HEX_06 ":ix_rkeys[" FMT_INT_04 "]:'%-8.8s'  ",
      |          ^~~~~~~~~~~~~~~~
      |          FMT_INT32_HEX
lldump.c:442:10: note: each undeclared identifier is reported only once for each function it appears in
lldump.c:442:26: error: expected ‘)’ before string constant
  442 |   printf(FMT_INT32_HEX_06 ":ix_rkeys[" FMT_INT_04 "]:'%-8.8s'  ",
      |                          ^~~~~~~~~~~~~
      |                          )
lldump.c:445:26: error: expected ‘)’ before string constant
  445 |   printf(FMT_INT32_HEX_06 ":ix_fkeys[" FMT_INT_04 "]:"
      |                          ^~~~~~~~~~~~~
      |                          )
lldump.c:473:29: error: expected ‘)’ before string constant
  473 |      printf(FMT_INT32_HEX_06 ":ix_rkeys[" FMT_INT_04 "]:'%-8.8s'  ",
      |                             ^~~~~~~~~~~~~
      |                             )
lldump.c:477:29: error: expected ‘)’ before string constant
  477 |      printf(FMT_INT32_HEX_06 ":ix_rkeys[" FMT_INT_04 "]:'%-8.8s'  ",
      |                             ^~~~~~~~~~~~~
      |                             )
lldump.c:480:28: error: expected ‘)’ before string constant
  480 |     printf(FMT_INT32_HEX_06 ":ix_fkeys[" FMT_INT_04 "]:"
      |                            ^~~~~~~~~~~~~
      |                            )
lldump.c: In function ‘print_block’:
lldump.c:729:18: error: expected ‘)’ before ‘FMT_INT_3’
  729 |     printf("    " FMT_INT_3 ". " FMT_INT32_HEX ":RKEY %8.8s "
      |                  ^~~~~~~~~~
      |                  )
lldump.c:757:18: error: expected ‘)’ before ‘FMT_INT_3’
  757 |     printf("    " FMT_INT_3 ". %-*s %s %s%s\n" ,
      |                  ^~~~~~~~~~
      |                  )
lldump.c:715:9: warning: unused variable ‘first’ [-Wunused-variable]
  715 |    char first = *akey.rkeyfirst;
      |         ^~~~~
lldump.c:795:18: error: expected ‘)’ before ‘FMT_INT_3’
  795 |     printf("    " FMT_INT_3 ". " FMT_INT32_HEX ":type '%c' "
      |                  ^~~~~~~~~~
      |                  )
lldump.c:794:11: warning: unused variable ‘offset2’ [-Wunused-variable]
  794 |     INT32 offset2 = (i+1)*8 + 4;
      |           ^~~~~~~
lldump.c:793:11: warning: unused variable ‘num’ [-Wunused-variable]
  793 |     INT32 num = *ptr++;
      |           ^~~
lldump.c:792:11: warning: unused variable ‘offset1’ [-Wunused-variable]
  792 |     INT32 offset1 = (i+1)*8;
      |           ^~~~~~~
lldump.c:791:11: warning: unused variable ‘type’ [-Wunused-variable]
  791 |     INT32 type = *ptr++;
      |           ^~~~
make[3]: *** [Makefile:470: lldump.o] Error 1
make[3]: Leaving directory '/home/parallels/git_stuff/lifelines/src/tools'
make[2]: *** [Makefile:345: all-recursive] Error 1
make[2]: Leaving directory '/home/parallels/git_stuff/lifelines/src'
make[1]: *** [Makefile:458: all-recursive] Error 1
make[1]: Leaving directory '/home/parallels/git_stuff/lifelines'
make: *** [Makefile:378: all] Error 2

The message is "*Bad history count"

I've downloaded the revised lldump.c file, but make failed, with the following forming part of the output:

gcc  -g -O2 -W -Wall -Wcast-align -Wmissing-declarations -Wmissing-prototypes -Wreturn-type -Wstrict-prototypes -pedantic -g   -o dbverify dbverify.o wprintf.o ../gedlib/libgedcom.a ../btree/libbtree.a ../stdlib/libstdlib.a ../arch/libarch.a  -lncursesw -lm 
gcc -DLOCALEDIR=\"/usr/local/share/locale\" -DHAVE_CONFIG_H -I. -I../..  -I../.. -I../../src/hdrs   -g -O2 -W -Wall -Wcast-align -Wmissing-declarations -Wmissing-prototypes -Wreturn-type -Wstrict-prototypes -pedantic -g -MT lldump.o -MD -MP -MF .deps/lldump.Tpo -c -o lldump.o lldump.c
lldump.c: In function ‘print_index’:
lldump.c:442:10: error: ‘FMT_INT32_HEX_06’ undeclared (first use in this function); did you mean ‘FMT_INT32_HEX’?
  442 |   printf(FMT_INT32_HEX_06 ":ix_rkeys[" FMT_INT_04 "]:'%-8.8s'  ",
      |          ^~~~~~~~~~~~~~~~
      |          FMT_INT32_HEX
lldump.c:442:10: note: each undeclared identifier is reported only once for each function it appears in
lldump.c:442:26: error: expected ‘)’ before string constant
  442 |   printf(FMT_INT32_HEX_06 ":ix_rkeys[" FMT_INT_04 "]:'%-8.8s'  ",
      |                          ^~~~~~~~~~~~~
      |                          )
lldump.c:445:26: error: expected ‘)’ before string constant
  445 |   printf(FMT_INT32_HEX_06 ":ix_fkeys[" FMT_INT_04 "]:"
      |                          ^~~~~~~~~~~~~
      |                          )
lldump.c:473:29: error: expected ‘)’ before string constant
  473 |      printf(FMT_INT32_HEX_06 ":ix_rkeys[" FMT_INT_04 "]:'%-8.8s'  ",
      |                             ^~~~~~~~~~~~~
      |                             )
lldump.c:477:29: error: expected ‘)’ before string constant
  477 |      printf(FMT_INT32_HEX_06 ":ix_rkeys[" FMT_INT_04 "]:'%-8.8s'  ",
      |                             ^~~~~~~~~~~~~
      |                             )
lldump.c:480:28: error: expected ‘)’ before string constant
  480 |     printf(FMT_INT32_HEX_06 ":ix_fkeys[" FMT_INT_04 "]:"
      |                            ^~~~~~~~~~~~~
      |                            )
lldump.c: In function ‘print_block’:
lldump.c:729:18: error: expected ‘)’ before ‘FMT_INT_3’
  729 |     printf("    " FMT_INT_3 ". " FMT_INT32_HEX ":RKEY %8.8s "
      |                  ^~~~~~~~~~
      |                  )
lldump.c:757:18: error: expected ‘)’ before ‘FMT_INT_3’
  757 |     printf("    " FMT_INT_3 ". %-*s %s %s%s\n" ,
      |                  ^~~~~~~~~~
      |                  )
lldump.c:715:9: warning: unused variable ‘first’ [-Wunused-variable]
  715 |    char first = *akey.rkeyfirst;
      |         ^~~~~
lldump.c:795:18: error: expected ‘)’ before ‘FMT_INT_3’
  795 |     printf("    " FMT_INT_3 ". " FMT_INT32_HEX ":type '%c' "
      |                  ^~~~~~~~~~
      |                  )
lldump.c:794:11: warning: unused variable ‘offset2’ [-Wunused-variable]
  794 |     INT32 offset2 = (i+1)*8 + 4;
      |           ^~~~~~~
lldump.c:793:11: warning: unused variable ‘num’ [-Wunused-variable]
  793 |     INT32 num = *ptr++;
      |           ^~~
lldump.c:792:11: warning: unused variable ‘offset1’ [-Wunused-variable]
  792 |     INT32 offset1 = (i+1)*8;
      |           ^~~~~~~
lldump.c:791:11: warning: unused variable ‘type’ [-Wunused-variable]
  791 |     INT32 type = *ptr++;
      |           ^~~~
make[3]: *** [Makefile:470: lldump.o] Error 1
make[3]: Leaving directory '/home/parallels/git_stuff/lifelines/src/tools'
make[2]: *** [Makefile:345: all-recursive] Error 1
make[2]: Leaving directory '/home/parallels/git_stuff/lifelines/src'
make[1]: *** [Makefile:458: all-recursive] Error 1
make[1]: Leaving directory '/home/parallels/git_stuff/lifelines'
make: *** [Makefile:378: all] Error 2

--- a/src/liflines/browse.c
+++ b/src/liflines/browse.c
@@ -1573,7 +1573,7 @@ load_nkey_list (STRING key, struct hist * histp)
 		return;
 	ptr = (INT32 *)rawrec;
 	temp = *ptr++;
-	if (temp<1 || temp > 9999) {
+	if (temp<0 || temp > 9999) {
 		/* #records failed sanity check */
 		msg_error("%s", _(qSbadhistcnt));
 		goto end;

--- a/src/liflines/browse.c
+++ b/src/liflines/browse.c
@@ -1561,15 +1561,17 @@ static void
 load_nkey_list (STRING key, struct hist * histp)
 {
 	STRING rawrec;
-	INT * ptr;
-	INT count, len, i, temp;
+	INT32 * ptr;
+	INT32 count;
+	INT32 temp;
+	INT len, i;
 
 	count = 0;
 	if (!(rawrec = retrieve_raw_record(key, &len)))
 		return;
 	if (len < 8 || (len % 8) != 0)
 		return;
-	ptr = (INT *)rawrec;
+	ptr = (INT32 *)rawrec;
 	temp = *ptr++;
 	if (temp<1 || temp > 9999) {
 		/* #records failed sanity check */
@@ -1635,7 +1637,9 @@ static void
 save_nkey_list (STRING key, struct hist * histp)
 {
 	FILE * fp=0;
-	INT next, count, temp;
+	INT i, next;
+	INT32 count;	// write buffer for histp->count value
+	INT32 temp;	// write buffer for histp->list[] values
 	size_t rtn;
 
 	count = get_hist_count(histp);
@@ -1649,9 +1653,12 @@ save_nkey_list (STRING key, struct hist * histp)
 	rtn = fwrite(&count, 4, 1, fp); ASSERT(rtn==1);
 	rtn = fwrite(&count, 4, 1, fp); ASSERT(rtn==1);
 
+	/* write entries */
 	next = histp->start;
-	while (1) {
+	for (i=0; i<count; ++i)
+	{
 		/* write type & number, 4 bytes each */
+		/* type = char, keynum = INT (truncated!) */
 		temp = histp->list[next].ntype;
 		rtn = fwrite(&temp, 4, 1, fp); ASSERT(rtn==1);
 		temp = histp->list[next].keynum;
@@ -2031,7 +2038,7 @@ get_vhist_len (void)
 	return get_hist_count(&vhist);
 }
 /*==================================================
- * get_vhist_len -- how many records currently in change history list ?
+ * get_chist_len -- how many records currently in change history list ?
  * Created: 2002/06/23, Perry Rapp
  *================================================*/
 INT

--- a/src/tools/lldump.c
+++ b/src/tools/lldump.c
@@ -250,6 +250,7 @@ print_usage (void)
  * 3c. N  6 chars long, N[A-Z$][A-Z][0-9]{3,3}
  *     Name
  * 3d. V is VUOPT (only V key I am aware of)
+ * 3e. H is HISTV or HISTC
  * messages to stdout - errors, warnings are printed and ignored
  * 
  *WARNING: item 2 is broken for Keys starting with a R, ie REFN data.
@@ -263,7 +264,7 @@ static void
 check_rkey(keytype *kt, RKEY *key,BOOLEAN in_data) {
 	if (*(INT64*)key == 0) {
 		// special case zero key
-		strncpy(kt->rkey,"8 x 0x00",9);  // null key in 8 chars
+		strncpy(kt->rkey,"0x00 x 8",9);  // null key in 8 chars
 		strncpy(kt->rname,"Zero",6);
 		kt->rkeyfirst = kt->rkey;
 		return;
@@ -292,6 +293,7 @@ check_rkey(keytype *kt, RKEY *key,BOOLEAN in_data) {
 	case 'S': strncpy(kt->rname,"SOUR",6);  break;
 	case 'V': strncpy(kt->rname,"VUOPT",6); break;
 	case 'X': strncpy(kt->rname,"Other",6); break;
+	case 'H': strncpy(kt->rname,"HIST",6);  break;
 	default:
 		printf("Error, unrecognized RKEY %s\n",p);
 		strncpy(kt->rname,"Bad",6); 
@@ -302,7 +304,8 @@ check_rkey(keytype *kt, RKEY *key,BOOLEAN in_data) {
 //    R,N are letter followed by A-Za-z0-9
 //       well really N is [A-Z]{3}[0-9]{3}
 //    REFN Name
-//    V is VUOPT`
+//    V is VUOPT
+//    H is HIST[VC]
 	switch (*p) {
 	case 'E': 
 	case 'F': 
@@ -348,6 +351,11 @@ check_rkey(keytype *kt, RKEY *key,BOOLEAN in_data) {
 		break;
 	case 'V': 
 		if (strcmp(p,"VUOPT") != 0) {
+			printf("Error, invalid letter in RKEY(%s)\n",kt->rkey);
+		}
+		break;
+	case 'H':
+		if (strcmp(p,"HISTV") != 0 && strcmp(p,"HISTC") != 0) {
 			printf("Error, invalid letter in RKEY(%s)\n",kt->rkey);
 		}
 		break;
@@ -692,7 +700,7 @@ void print_block(BTREE btree, BLOCK block, INT32 *offset)
 			}
 		}
 		check_rkey(&akey,&rkeys(block,n),TRUE);
-		/* keys start with I,F,S,E,X and N R V */
+		/* keys start with I,F,S,E,X and N R V H */
 		printf("[" FMT_INT_04 "] %s rkey: %s offs: " FMT_INT32_HEX
 			" lens: " FMT_INT32_HEX "\n",
 			n,akey.rname, akey.rkeyfirst,
@@ -757,6 +765,40 @@ void print_block(BTREE btree, BLOCK block, INT32 *offset)
 				lennames += strlen(&rec[stroff])+1;
 				col1 += sizeof(RKEY);
 				col2 += sizeof(INT32);
+			}
+		} else if (*akey.rkeyfirst == 'H') {
+			// HIST data consists of
+			// INT16 count : number of history records
+			// INT16 count : number of history records (again)
+			// .. foreach ..
+			// INT16 ntype : node type (from char)
+			// INT16 nkey  : node key (from INT)
+			//
+			// See save_nkey_list and load_nkey_list
+			//
+			INT32 *ptr = &rec[0];
+			INT32 count1 = *ptr++;
+			INT32 count2 = *ptr++;
+			INT32 count = ((count1>count2) ? count2 : count1);
+			INT i;
+
+			// print out count
+			printf("   " FMT_INT32_HEX ": count1 " FMT_INT32 "\n",
+				*offset,count1);
+			printf("   " FMT_INT32_HEX ": count1 " FMT_INT32 "\n",
+				*offset+sizeof(INT32),count2);
+
+			// print out entries
+			for (i=0; i<count; i++)
+			{
+				INT32 type = *ptr++;
+				INT32 offset1 = (i+1)*8;
+				INT32 num = *ptr++;
+				INT32 offset2 = (i+1)*8 + 4;
+				printf("    " FMT_INT_3 ". " FMT_INT32_HEX ":type '%c' "
+				                             FMT_INT32_HEX ":keynum " FMT_INT "\n",
+					i, *offset + offset1, (char)type,
+					   *offset + offset2, (INT)num);
 			}
 		} else {
 			// handle all but N,R - E F I P S V X 

--- a/src/tools/lldump.c
+++ b/src/tools/lldump.c
@@ -760,7 +760,7 @@ void print_block(BTREE btree, BLOCK block, INT32 *offset)
 				printf("    " FMT_INT_3 ". %-*s %s %s%s\n" ,
 					m+1, keylen, akey.rkeyfirst,
 					(first == 'N' ? " name " :
-				  	" identified by 'REFN "), &rec[stroff],
+					" identified by 'REFN "), &rec[stroff],
 					(first == 'N' ? "" : "'"));
 				lennames += strlen(&rec[stroff])+1;
 				col1 += sizeof(RKEY);
@@ -776,7 +776,7 @@ void print_block(BTREE btree, BLOCK block, INT32 *offset)
 			//
 			// See save_nkey_list and load_nkey_list
 			//
-			INT32 *ptr = &rec[0];
+			INT32 *ptr = (INT32 *)&rec[0];
 			INT32 count1 = *ptr++;
 			INT32 count2 = *ptr++;
 			INT32 count = ((count1>count2) ? count2 : count1);
@@ -786,7 +786,7 @@ void print_block(BTREE btree, BLOCK block, INT32 *offset)
 			printf("   " FMT_INT32_HEX ": count1 " FMT_INT32 "\n",
 				*offset,count1);
 			printf("   " FMT_INT32_HEX ": count1 " FMT_INT32 "\n",
-				*offset+sizeof(INT32),count2);
+				*offset+(INT32)sizeof(INT32),count2);
 
 			// print out entries
 			for (i=0; i<count; i++)
@@ -801,7 +801,7 @@ void print_block(BTREE btree, BLOCK block, INT32 *offset)
 					   *offset + offset2, (INT)num);
 			}
 		} else {
-			// handle all but N,R - E F I P S V X 
+			// handle all but N,R,H - E F I P S V X
 			// these have just text data
 			if (rec != NULL)
 				printf(FMT_INT32_HEX "-" FMT_INT32_HEX 

